release/0.2.x candidate (#1126)
GEOSX/GEOSX: src/docs/sphinx/developerGuide/Contributing/GitWorkflow.rst

of ``develop`` into a ``feature`` branch.
The naming of feature branches should follow ``feature/[developer]/[branch-description]``
if you expect that only a single developer will contribute to the branch, 
or ``feature/[branch-description]`` if you expect it will be a collaborative effort.
For example, if a developer named ``neo`` were to add or modify a code feature
expecting that he/she would be the only contributor, he/she would create a branch 
using the following commands to create the local branch and push it to the remote
repository:

.. code-block:: sh

  git checkout -b feature/neo/freeYourMind
  git push -u origin feature/neo/freeYourMind

However if the branch is a collaborative branch amongst many developers, the 
appropriate commands would be:

.. code-block:: sh

  git checkout -b feature/freeYourMind
  git push -u origin feature/freeYourMind
  
When ``feature`` branches are ready to be merged into ``develop``, a ``Pull Request``
should be created to perform the review and merging process. 
See below for details about :ref:`Submitting_a_Pull_Request`.

Bugfix Branches
---------------
Bugfix branches are used to fix bugs that are present in the ``develop`` branch.

> A similar naming convention to that of the ``feature`` branches are used, replacing

Author: ```suggestion
A similar naming convention to that of the ``feature`` branches is used, replacing
```

development of ``feature/smith/dodgeBullets``.
This would look like:

.. code-block:: sh
   git checkout develop
   git pull
   git checkout feature/smith/dodgeBullets
   git rebase -onto G CC

As should be apparent, we have specified the starting point as ``G``, and the point 
at which we replay the commits in ``feature/smith/dodgeBullets`` as all commits 
AFTER ``CC``.
The result is:

.. code-block:: sh

   A----B----C----D----E----F----G----E (develop)
                                  \
                                 CCA'--CCB'--CCC' (feature/smith/dodgeBullets)

Now you may proceed with standard methods for keeping ``feature/smith/dodgeBullets`` 
current with ``develop``.

.. _Submitting_a_Pull_Request:

Submitting a Pull Request
======================================
Once you have created your branch and pushed to Github, you can now create a 
`Pull Request <https://github.com/GEOSX/GEOSX/pulls>`_ on Github.
The PR creates a central place to review and discuss the ongoing work on the branch. 
Creating a pull request early in the development process is preferred as it allows 
for developers to collaborate on the branch more readily.

.. note::

>    When initially creating a pull request (PR) on GitHub, create it is a draft PR while

Reviewer: ```suggestion
   When initially creating a pull request (PR) on GitHub, create it as a draft PR while
```

If it fails (perhaps because of your institution's proxy),
you may consider the `personal access token option <https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line>`_ as an alternative.

> Downloading The Code

Reviewer 2: ```suggestion
Downloading the Code
```
(words of three letters or less not capitalized)

If all goes well, you should have a complete copy of the GEOSX source at this point.
The most common errors people encounter here have to do with Github not recognizing
thier authentication settings.

Branching Model
===============
The current branching model used in GEOSX (starting with version 0.2) is essentially the 
`Oneflow branching model and workflow <https://www.endoflineblog.com/oneflow-a-git-branching-model-and-workflow#when-not-to-use-oneflow/>`_
with some modifications to the merging strategy, and the treatment of release 
branches. 
In GEOSX, the ``develop`` branch serves as the main branch in the ``Oneflow`` model.
The remaining branch types are described in the following subsections.

.. note::
   The early commits in GEOSX (up to version 0.2) used a 
   `Gitflow <https://nvie.com/posts/a-successful-git-branching-model/>`_
   approach for merging feature branches into develop.
   This was done without cleaning the commit history in each feature
   branch prior to the merge into develop, resulting in an overly verbose 
   (i.e. useless) history.
   Furthermore, as would be expected, having many active feature branches resulted
   in a fairly wide/spaghetti history.

>    As some point in the development process, we choose try switch primarily to a 

Reviewer 2: ```suggestion
   As some point in the development process, we chose to switch primarily to a 
```

prior to merging into ``develop``.

Keeping Your Branch Current
===========================
Over the course of a long development effort in a single ``feature`` branch, a 
developer may need either merge ``develop`` into their ``feature`` branch, or rebase
their ``feature`` branch on ``develop``.
We don't have a mandate on how you keep your branch current, but we do have 
guidelines on the branch history when merging your branch into ``develop``.
Typically, merging ``develop`` into your branch is the easiest approach, but will
lead to a complex relationship with ``develop`` with multiple interactions...which 
can lead to a confusing history.
Conversely, rebasing your branch onto ``develop`` is more difficult, but will lead 
to a linear history within the branch.
For a complex history, we will perform a squash merge into ``develop``, thereby 
the work from the branch will appear as a single commit in ``develop``.
For clean branch histories where the individual commits are meaningful and should 
be preserved, we have to option to perform a merge commit in with the PR is merged 
into ``develop``, with the addition of a merge commit, thus maintaining the commit history.

> Branching Off of a Branch

Reviewer 2: ```suggestion
Branching off of a Branch
```

current with ``develop``.

.. _Submitting_a_Pull_Request:

Submitting a Pull Request
======================================
Once you have created your branch and pushed to Github, you can now create a 
`Pull Request <https://github.com/GEOSX/GEOSX/pulls>`_ on Github.
The PR creates a central place to review and discuss the ongoing work on the branch. 
Creating a pull request early in the development process is preferred as it allows 
for developers to collaborate on the branch more readily.

.. note::
   When initially creating a pull request (PR) on GitHub, create it as a draft PR while
   work is ongoing and the PR isn't ready for testing, review, and merge consideration.

When you create the initial draft PR, please ensure that you apply appropriate labels.
Applying labels allows other developers to more quickly filter the live PRs and access
those that are relevant to them. Always add the `new` label upon PR creation, as well
as to the appropriate `type`, `priority`, and  `effort` labels. In addition, please
also add any appropriate `flags`.

.. note::
   If your branch and PR will resolve any open issues, be sure to `link` them to

>    the PR to ensure they are appropriate resolved once the PR is merged.

Reviewer 3: ```suggestion
   the PR to ensure they are appropriately resolved once the PR is merged.
```
